vtk: update package using changed variable names

glennpj opened the pull request:
This PR updates the vtk package to use new variable names and adds some
external dependencies.

- add version 1.4.2 to gl2ps package. This is needed to use gl2ps as a
  dependency.
- new package: utf8. This can now be sued as a dependency for version 9.
- use the package provided FindHDF5.cmake for versions up to 8. Version
  9 already uses an internal version.
- adjust the hdf dependencies to match the ~mpi and +mpi variants
- add external dependencies when possible
    - pugixml
    - libogg
    - libtheora
    - utf8
    - gl2ps
    - proj
- turn off configuring MPI if ~mpi
- use new variables for version 9
- remove unused variables depending on version

glennpj committed:
vtk: update package using changed variable names

This PR updates the vtk package to use new variable names and adds some
external dependencies.

- add version 9.1.0
- add version 1.4.2 to gl2ps package. This is needed to use gl2ps as a
  dependency.
- new package: utf8. This can now be sued as a dependency for version 9.
- use the package provided FindHDF5.cmake for versions up to 8. Version
  9 already uses an internal version.
- adjust the hdf dependencies to match the ~mpi and +mpi variants
- add external dependencies when possible
    - pugixml
    - libogg
    - libtheora
    - utf8
    - gl2ps
    - proj
- turn off configuring MPI if ~mpi
- use new variables for version 9
- remove unused variables depending on version
- make seacas work yywith ~mpi

scheibelp commented on var/spack/repos/builtin/packages/seacas/package.py:
```diff
@@ -86,7 +86,7 @@ class Seacas(CMakePackage):
 
     depends_on('netcdf-c@4.8.0:+mpi+parallel-netcdf', when='+mpi')
-    depends_on('netcdf-c@4.8.0:~mpi', when='~mpi')
+    depends_on('netcdf-c@4.8.0:~mpi~parallel-netcdf', when='~mpi')
```
Question: is it not possible to use `netcdf-c+parallel-netcdf` with `~mpi`? I'm curious if we can still use the dependency built with `+parallel-netcdf` even if we don't want to use `+mpi` for `seacas`.

glennpj replied:
That is a good question and I think the answer is, "it is possible", because `parallel-netcdf`  does not depend on `netcdf-c`. However, the `netcdf-c` package recipe has the following:
```
 depends_on('mpi', when='+parallel-netcdf')
```
which makes the above dependency necessary in `seacas`. I was not inclined to mess with the `netcdf-c` packages in this PR.

glennpj replied:
I reverted this change as the clingo concretizer handles this natively.

scheibelp commented on var/spack/repos/builtin/packages/utf8/package.py:
```diff
@@ -0,0 +1,20 @@
+from spack import *
+
+
+class Utf8(Package):
```
Could this be updated to use the full package name: `Utf8Cpp` which should be installable like `spack install utf8-cpp`

glennpj replied:
Yes, it could be, I even had it that way at one point while working on this. I settled on `utf8` as that is what the zip file is called. The `spack create` command determined the name was `utf` so this one is odd.

glennpj replied:
@scheibelp Can you clarify whether this should be `utf8-cpp` or `utf8cpp`? I ask because the naming for this package is all over the map. The project homepage refers to `utf8-cpp`. Other than that, the code for the software is referred to as:

- utfcpp
- utf8cpp
- utf8

The VTK project seems to prefer `utf8cpp` so that is what I would lean towards.

scheibelp replied:
I prefer `utf8-cpp` (to achieve that, you would choose `Utf8Cpp` as the name of the Package class in the `package.py` file).

scheibelp commented on var/spack/repos/builtin/packages/vtk/package.py:
```diff
@@ -101,7 +105,8 @@ class Vtk(CMakePackage):
     # set hl variant explicitly, similar to issue #7145
-    depends_on('hdf5+hl')
+    depends_on('hdf5+hl~mpi', when='~mpi')
+    depends_on('hdf5+hl+mpi', when='+mpi')
```
Does this definitely require `hdf5+mpi` when `+mpi`? I've encountered packages that build without `mpi` but can still use `hdf5+mpi`.

glennpj replied:
This is similar to the previous question about `mpi`. I set it as above to be consistent with how the `netcdf-c` package handles the `mpi` variant. However, I am pretty sure that I was still using the original concretizer when I developed this PR so the extra consistency helped reduce conflicts. With the clingo concretizer this is probably no longer relevant. So, reverting this change should be okay.

glennpj replied:
I reverted this as well as the `netcdf` change in `seacas`. I tested with the `clingo` concretizer and as suspected, it resolves everything to `~mpi`.

scheibelp commented on var/spack/repos/builtin/packages/vtk/package.py:
```diff
@@ -218,14 +247,14 @@ def cmake_args(self):
-                '-DQT_QMAKE_EXECUTABLE:PATH={0}'.format(qmake_exe),
-                '-DVTK_Group_Qt:BOOL=ON',
-            ])
             # https://github.com/martijnkoopman/Qt-VTK-viewer/blob/master/doc/Build-VTK.md
```
It looks like this comment is supposed to be attached to the `else` now.

glennpj replied:
No, the document that the link refers to has been updated over time for newer versions, at one point it referred to VTK-8. Perhaps I should add a comment to note that.

glennpj replied:
I added a comment for this in the package.

scheibelp commented on var/spack/repos/builtin/packages/vtk/package.py:
```diff
@@ -147,38 +164,45 @@ def cmake_args(self):
         else:
             cmake_args.extend([
-                '-DVTK_USE_SYSTEM_OGGTHEORA:BOOL=OFF',
+                '-DVTK_USE_EXTERNAL:BOOL=ON',
```
(question: no change needed) Did this replace `VTK_USE_SYSTEM_LIBRARIES`?

glennpj replied:
Yes.

scheibelp commented on var/spack/repos/builtin/packages/vtk/package.py:
```diff
@@ -235,37 +264,54 @@ def cmake_args(self):
+                    "-DVTK_USE_SYSTEM_XDMF3:BOOL=OFF",
+                    "-DVTK_USE_SYSTEM_XDMF2:BOOL=OFF"
+                ])
+            else:
```
I see `HDF5_ROOT` is handled elsewhere here but not `BOOST_ROOT` - is that no-longer necessary for `@9:`?

glennpj replied:
Correct, `@9:` is able to find `boost`.

scheibelp reviewed (changes requested):
I have a few requests and a few questions.

glennpj committed:
Incorporate suggested changes

- reverted netcdf dep in seacas
- reverted hdf dep in vtk
- changed utf to utf8cpp
  - add url_for_version

scheibelp commented on var/spack/repos/builtin/packages/seacas/package.py:
```diff
@@ -126,6 +126,8 @@ def cmake_args(self):
                 '-DMPI_BASE_DIR:PATH=%s'      % spec['mpi'].prefix,
             ])
+        else:
+            '-DTPL_ENABLE_MPI:BOOL=OFF',
```
This trailing comma looks odd to me but I wouldn't be shocked if it is somehow valid Python - have you successfully installed with `~mpi` since making this change?

glennpj replied:
Yes, I did build it successfully. However, the stray comma is likely a
"copy --> paste --> (in)complete edit" error and does not belong there. I just added a commit to fix it.

glennpj committed:
Remove stray comma

scheibelp commented:
I was reading through this and noticed:

> use the package provided FindHDF5.cmake for versions up to 8. Version 9 already uses an internal version.

It looks like this PR sets `HDF5_ROOT` for `@9:`, i.e. that it uses a Spack-built HDF5 - that's a good thing IMO (generally we don't want a package to install its own dependencies) but I wanted to double check with you because that contradicts this line from the commit message.

glennpj commented:
> I was reading through this and noticed:
> 
> > use the package provided FindHDF5.cmake for versions up to 8. Version 9 already uses an internal version.
> 
> It looks like this PR sets `HDF5_ROOT` for `@9:`, i.e. that it uses a Spack-built HDF5 - that's a good thing IMO (generally we don't want a package to install its own dependencies) but I wanted to double check with you because that contradicts this line from the commit message.

There is no contradiction, just unclear wording on my part due to lacking context. The comment is in reference to the cmake files/code used to find HDF5, not an internal implementation of HDF5. While there is a cmake file that is part of the cmake distribution for finding HDF5, it does not work well with the HL component. This is a known issue and VTK includes its own custom cmake code for finding `hdf5+hl` to address this. For `vtk@:8` however, it will try to use cmake's distributed code to find `hdf5+hl` for newer cmake versions, but it will fail to find `hdf5+hl`. I included a patch so `vtk@:8` will always use its own, older, cmake code rather than that distributed with cmake. Now, `vtk@9:` always uses its own cmake code so a patch is not needed. The respective `vtk@:8` and `vtk@9:` cmake codes use different methods to find `hdf5+hl`, and the `vtk@9:` code needs `HDF5_ROOT` to be set, while the `vtk@:8` code does not.

In summary:
- `vtk@:8` > needs a patch to use vtk provided cmake code but does not need `HDF5_ROOT` set
- `vtk@9:` > does not need a patch to use vtk provided cmake code but needs `HDF5_ROOT` set

scheibelp commented:
@spackbot re-run pipeline
